rfd/0007-rbac-oss.md

```diff
@@ -0,0 +1,160 @@
+---
+authors: Alexander Klizhentas (sasha@gravitational.com)
+state: discussion
+---
+
+# RFD 2 - Open Source Roles
+
+## What
+
+Enable role based access control in open source.
+
+## Why
+
+Provide better user experience for open source users.
+Unify user experience, testing and development for open source branches.
+
+## Details
+
+### New Open Source Features
+
+The following features will become available starting 5.1 version.
+
+**Role Based Access Control**
+
+All RBAC features, except FedRamp feature flags.
+
+**Approval Plugins**
+
+Some approval plugins will become available in the open source:
+
+* Approval Workflows Golang SDK and API
+* Webhook
+* Slack
+* Gitlab
+* Mattermost
+
+### Enteprise Only Features
+
+The following features will remain enterprise only.
+
+**FedRamp**
+
+* Max connections AC control (and all future AC controls):
+
+```yaml
+role:
+   options:
+     max_connections: 3
+```
+
+* Teleport fips mode flag
+
+```bash
+$ teleport start --fips
+```
+
+**Single Sign On**
+
+OIDC and SAML connectors
+
+**Approval Plugins**
+
+Some plugins will be require license to start
+and will be moved to private source repository.
+
+* JIRA Plugin
+* PagerDuty Plugin
+
+**Extended approval workflows**
+
+New feature flags allowing waiting room and always requesting flow
+will be enterprise only. 
+
+```yaml
+role:
+  options:
+     # enterprise only values
+     request_access: 'note|always'
+```
+
+OSS users can request roles using `tsh login --request-roles` on demand.
+
+User interface with approval requests, waiting room remains in enterprise.
+
+### Migration Details
+
+**Open Source Users**
+
+Open source users will be assigned to a new `user` role. This role
+is almost backwards compatible with builtin OSS role `admin`,
+except it does not allow to modify resources. Otherwise all users
+will become admins after migration:
+
+```yaml
+role:
+   name: user
+spec:
+  options:
+    port_forwarding: true
+    max_session_ttl: 30h
+    forward_agent: true
+    enhanced_recording: ['command', 'network']
+  allow:
+    logins: ['{{internal.logins}}']
+    node_labels: '*': '*'
+```
+
+Another role, `admin` will be created.
+
+```yaml
+kind: role
+metadata:
+  name: admin
+spec:
+  allow:
+    logins: ['this-login-does-not-exist']
+    rules:
+    - resources: ['*']
+      verbs: ['*']
+  deny: {}
+```
+
+Migration tutorial will advice to promote designated local user
+to admin:
+
+```bash
+$ tctl users update alice --set-roles=admin
+```
+
+**Github Connector**
+
+Github in open source mode will support both `teams_to_logins`
+and `teams_to_roles` modes.
+
+**Adding users in tctl**
+
+Both OSS (legacy) user add will be supported to preserve backwards
+compatibility.
+
+```
+# Adding a user to Teleport with the principle joe, root & ec2-user
+$ tctl users add joe joe,root,e2-user
+
+# Becomes alias of
+$ tctl users add joe --traits=internal.logins=joe,root,e2-user --roles=user
```

Comment:
We should heavily discourage the use of this - perhaps print a warning when someone uses this command and say "hey, you should use RBAC instead"

Reason: we test our RBAC flows much more often.